File: LeetCode/454-4sum-ii/ex_454.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        map<int, int> sum_map;
        int count = 0;
        for (auto &i : A) {
            for (auto &j :B) {
                ++sum_map[i + j];
            }
        }
        
        for (auto &i : C) {
            for (auto &j :D) {
                if (sum_map.find(-i - j) != sum_map.end()) {
                    count += sum_map[-i - j];
                }
            }
        }
        return count;
    }
};

File: LeetCode/454-4sum-ii/ex_454.java
public class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        HashMap<Integer, Integer> sumMap = new HashMap<>();
        int count = 0;
        
        for (int i : A) {
            for (int j : B) {
                sumMap.put(i + j, sumMap.getOrDefault(i + j, 0) + 1);
            }
        }
        
        for (int i : C) {
            for (int j : D) {
                if (sumMap.containsKey(-i - j)) {
                    count += sumMap.get(-i - j);
                }
            }
        }
        
        return count;
    }
}